lib/hypothesis/providers.rb

```diff
@@ -2,24 +2,90 @@
 
 module Hypothesis
   module Providers
+    def bits(n)
+      from_hypothesis_core HypothesisCoreBitProvider.new(n)
+    end
+
+    def composite(&block)
+      Hypothesis::Provider::Implementations::CompositeProvider.new(self, block)
+    end
+
     def integers
-      proc do |source|
-        if source.bits(1).positive?
-          source.bits(64)
+      composite do
+        if given(bits(1)).positive?
```

Comment:
If I understand this right, `given` triggers `local_given` which goes to the provider, which in a composite block is an `instance_eval` of a new composite object. Then we do this whole block again, just under the new composite object, which has `method_missing` for `bits`, so it goes to the parent and then Helix. Is this the correct interpretation?

Reply:
I think that's right but I'm not totally following your notation so let me spell it out:

* `x = self.bits(1) = @parent.bits(1)
* `self.given(x) -> @source.local_given(x)`
* `@source.local_given(x) -> x.provide(@source)`
* `x.provide(@source) -> x.@core_provider.provide(@source.wrapped_data)` (which is a Helix method)